src/besselj.jl

```diff
@@ -1,42 +1,77 @@
-#=
-Cephes Math Library Release 2.8:  June, 2000
-Copyright 1984, 1987, 2000 by Stephen L. Moshier
-https://github.com/jeremybarnes/cephes/blob/master/bessel/j0.c
-https://github.com/jeremybarnes/cephes/blob/master/bessel/j1.c
-=#
-function besselj0(x::Float64)
-    T = Float64
+#    Bessel functions of the first kind of order zero and one
+#                       besselj0, besselj1
+#
+#    Calculation of besselj0 is done in three branches using polynomial approximations
+#
+#    Branch 1: x <= 5.0
+#              besselj0 = (x^2 - r1^2)*(x^2 - r2^2)*P3(x^2) / Q8(x^2)
+#    where r1 and r2 are zeros of J0
+#    and P3 and Q8 are a 3 and 8 degree polynomial respectively
+#    Polynomial coefficients are from [1] which is based on [2]
+#    For tiny arugments the power series expansion is used.
+#
+#    Branch 2: 5.0 < x < 75.0
+#              besselj0 = sqrt(2/(pi*x))*(cos(x - pi/4)*R7(x) - sin(x - pi/4)*R8(x))
+#    Hankel's asymptotic expansion is used
+#    where R7 and R8 are rational functions (Pn(x)/Qn(x)) of degree 7 and 8 respectively
+#    See section 4 of [3] for more details and [1] for coefficients of polynomials
+# 
+#   Branch 3: x >= 75.0
+#              besselj0 = sqrt(2/(pi*x))*beta(x)*(cos(x - pi/4 - alpha(x))
+#   See modified expansions given in [3]. Exact coefficients are used
+#
+#   Calculation of besselj1 is done in a similar way as besselj0.
+#   See [3] for details on similarities.
+# 
+# [1] https://github.com/deepmind/torch-cephes
+# [2] Cephes Math Library Release 2.8:  June, 2000 by Stephen L. Moshier
+# [3] Harrison, John. "Fast and accurate Bessel function computation." 
+#     2009 19th IEEE Symposium on Computer Arithmetic. IEEE, 2009.
+#
+function besselj0(x::T) where T
     x = abs(x)
-    iszero(x) && return one(x)
     isinf(x) && return zero(x)
 
     if x <= 5
         z = x * x
         if x < 1.0e-5
             return 1.0 - z / 4.0
         end
-
-        DR1 = 5.78318596294678452118E0
-        DR2 = 3.04712623436620863991E1
-
+        DR1 = 5.78318596294678452118e0
+        DR2 = 3.04712623436620863991e1
         p = (z - DR1) * (z - DR2)
         p = p * evalpoly(z, RP_j0(T)) / evalpoly(z, RQ_j0(T))
         return p
-    else
+    elseif x < 75.0
         w = 5.0 / x
         q = 25.0 / (x * x)
 
         p = evalpoly(q, PP_j0(T)) / evalpoly(q, PQ_j0(T))
         q = evalpoly(q, QP_j0(T)) / evalpoly(q, QQ_j0(T))
         xn = x - PIO4(T)
-        p = p * cos(xn) - w * q * sin(xn)
+        sc = sincos(xn)
+        p = p * sc[2] - w * q * sc[1]
         return p * SQ2OPI(T) / sqrt(x)
+    else
+        xinv = inv(x)
+        x2 = xinv*xinv
+
+        p = (one(T), -1/16, 53/512, -4447/8192, 5066403/524288)
+        p = evalpoly(x2, p)
+        a = SQ2OPI(T) * sqrt(xinv) * p
+
+        q = (-1/8, 25/384, -1073/5120, 375733/229376, -55384775/2359296)
+        xn = muladd(xinv, evalpoly(x2, q), - PIO4(T))
```

Comment:
This subtraction of pi/4 would be better handled in `cos_sum` since we could just subtract 1 from `n` to shift the quadrant by 1.

Reply:
Yes definitely better to do it that way!